File: listas/src/main/java/com/victorlh/android/framework/listas/adapter/seleccionable/ListaSeleccionableAdapter.java
package com.victorlh.android.framework.listas.adapter.seleccionable;

import android.util.Log;

import androidx.annotation.NonNull;

import com.victorlh.android.framework.listas.ItemLista;
import com.victorlh.android.framework.listas.adapter.AbstractLista;
import com.victorlh.android.framework.listas.adapter.AbstractViewHolder;
import com.victorlh.android.framework.listas.adapter.ListaObservable;

import java.lang.reflect.Array;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

public class ListaSeleccionableAdapter<T extends ItemLista> implements ListaSeleccionable<T> {

	private AbstractLista<T> listaAdapter;
	private ListaSeleccionableObservable observable;

	private ETipoSeleccion tipoSeleccion;
	private List<T> seleccion;

	private List<OnStartSeleccionListener> onStartSeleccionListener;
	private List<OnStopSeleccionListener> onStopSeleccionListener;
	private List<OnSeleccionChangeListener> onSeleccionChangeListeners;
	private List<OnItemSeleccionadoListener<T>> onItemSeleccionadoListeners;
	private List<OnItemDeseleccionadoListener<T>> onItemDeseleccionadoListeners;

	public ListaSeleccionableAdapter(ETipoSeleccion tipoSeleccion) {
		this.tipoSeleccion = tipoSeleccion;

		this.seleccion = new ArrayList<>();
		this.onStartSeleccionListener = new ArrayList<>();
		this.onStopSeleccionListener = new ArrayList<>();
		this.onSeleccionChangeListeners = new ArrayList<>();
		this.onItemSeleccionadoListeners = new ArrayList<>();
		this.onItemDeseleccionadoListeners = new ArrayList<>();

		this.observable = new ListaSeleccionableObservable();
	}

	@Override
	public void setTipoSeleccion(@NonNull ETipoSeleccion tipoSeleccion) {
		this.tipoSeleccion = tipoSeleccion;
		if (tipoSeleccion == ETipoSeleccion.NONE && haySeleccionados()) {
			removeAllSeleccion();
		}
	}

	@Override
	public ETipoSeleccion getTipoSeleccion() {
		return tipoSeleccion;
	}

	@Override
	public void setSeleccion(@NonNull T[] items) {
		setSeleccion(Arrays.asList(items));
	}

	@Override
	public void setSeleccion(@NonNull List<T> items) {
		checkAttached();
		removeAllSeleccion();
		for (T item : items) {
			long id = item.getId();
			T elementoLista = listaAdapter.getItemById(id);
			if (elementoLista != null) {
				addSeleccion(elementoLista);
			}
		}
	}

	@Override
	public void addSeleccion(@NonNull T item) {
		checkAttached();
		boolean haySeleccionados = haySeleccionados();
		seleccion.add(item);
		AbstractViewHolder<T> viewHolder = listaAdapter.getViewHolder(item);
		if (viewHolder != null) {
			viewHolder.setSeleccion(true);
		}
		if (!haySeleccionados) {
			startSeleccion();
		}
		onItemSeleccionado(item);
		onSeleccionChange();
	}

	@Override
	public void removeSeleccion(@NonNull T item) {
		Log.d("AAAAAA", "Remove seleccion");
		checkAttached();
		seleccion.remove(item);
		AbstractViewHolder<T> viewHolder = listaAdapter.getViewHolder(item);
		if (viewHolder != null) {
			viewHolder.setSeleccion(false);
		}
		onItemDeseleccionado(item);
		onSeleccionChange();
		if (!haySeleccionados()) {
			stopSeleccion();
		}
	}

	@Override
	public void removeAllSeleccion() {
		checkAttached();
		seleccion.clear();
		List<T> lista = listaAdapter.getLista();
		for (T item : lista) {
			AbstractViewHolder<T> viewHolder = listaAdapter.getViewHolder(item);
			if (viewHolder != null) {
				viewHolder.setSeleccion(false);
			}
			onItemDeseleccionado(item);
		}
		onSeleccionChange();
		stopSeleccion();
	}

	private void removeAllSeleccionSinEventos() {
		checkAttached();
		seleccion.clear();
		List<T> lista = listaAdapter.getLista();
		for (T item : lista) {
			AbstractViewHolder<T> viewHolder = listaAdapter.getViewHolder(item);
			if (viewHolder != null) {
				viewHolder.setSeleccion(false);
			}
		}
	}

	@Override
	public boolean haySeleccionados() {
		checkAttached();
		return !seleccion.isEmpty();
	}

	@Override
	public boolean isSeleccionado(@NonNull T item) {
		checkAttached();
		return seleccion.contains(item);
	}

	@Override
	@SuppressWarnings("unchecked")
	public T[] getItemsSeleccionados() {
		checkAttached();
		Class<T> clazz = listaAdapter.getClazz();
		return seleccion.toArray((T[]) Array.newInstance(clazz, 0));
	}

	@Override
	public void addOnStartSeleccionListener(@NonNull OnStartSeleccionListener onStartSeleccionListener) {
		this.onStartSeleccionListener.add(onStartSeleccionListener);
	}

	@Override
	public void addOnStopSeleccionListener(@NonNull OnStopSeleccionListener onStopSeleccionListener) {
		this.onStopSeleccionListener.add(onStopSeleccionListener);
	}

	@Override
	public void addOnSeleccionChangeListener(@NonNull OnSeleccionChangeListener onSeleccionChangeListener) {
		this.onSeleccionChangeListeners.add(onSeleccionChangeListener);
	}

	@Override
	public void addOnItemDeseleccionadoListener(@NonNull OnItemDeseleccionadoListener<T> onItemDeseleccionadoListener) {
		this.onItemDeseleccionadoListeners.add(onItemDeseleccionadoListener);
	}

	@Override
	public void addOnItemSeleccionadoListener(@NonNull OnItemSeleccionadoListener<T> onItemSeleccionadoListener) {
		this.onItemSeleccionadoListeners.add(onItemSeleccionadoListener);
	}

	public void attachLista(AbstractLista<T> lista) {
		this.listaAdapter = lista;
		this.listaAdapter.registrarObservable(observable);
	}

	public void detachLista() {
		this.listaAdapter.removerObservable(observable);
		this.listaAdapter = null;
	}

	private void startSeleccion() {
		for (int i = 0; i < onStartSeleccionListener.size(); i++) {
			OnStartSeleccionListener listener = this.onStartSeleccionListener.get(i);
			listener.onStartSeleccion();
		}
	}

	private void stopSeleccion() {
		for (int i = 0; i < onStopSeleccionListener.size(); i++) {
			OnStopSeleccionListener listener = this.onStopSeleccionListener.get(i);
			listener.onStopSeleccion();
		}
	}

	private void onSeleccionChange() {
		int size = seleccion.size();
		for (int i = 0; i < onSeleccionChangeListeners.size(); i++) {
			OnSeleccionChangeListener listener = this.onSeleccionChangeListeners.get(i);
			listener.onSeleccionChangeListener(size);
		}
	}

	private void onItemSeleccionado(T item) {
		for (OnItemSeleccionadoListener<T> onItemSeleccionadoListener : onItemSeleccionadoListeners) {
			onItemSeleccionadoListener.onItemSeleccionado(item);
		}
	}

	private void onItemDeseleccionado(T item) {
		for (OnItemDeseleccionadoListener<T> onItemDeseleccionadoListener : onItemDeseleccionadoListeners) {
			onItemDeseleccionadoListener.onItemDeseleccionado(item);
		}
	}

	private void checkAttached() {
		if (listaAdapter == null) {
			throw new IllegalStateException("No se a asociado el adapter a ninguna Lista");
		}
	}

	private class ListaSeleccionableObservable extends ListaObservable<T> {

		@Override
		protected void recargarLista() {
			removeAllSeleccion();
		}

		@Override
		protected boolean onClickElemento(AbstractViewHolder<T> viewHolder) {
			T item = viewHolder.getItem();
			boolean haySeleccionados = haySeleccionados();
			if (tipoSeleccion != ETipoSeleccion.NONE) {
				if (!haySeleccionados) {
					if (tipoSeleccion != ETipoSeleccion.LONGCLICK) {
						addSeleccion(item);
						return true;
					}
				} else {
					if (isSeleccionado(item)) {
						removeSeleccion(item);
					} else {
						if (tipoSeleccion == ETipoSeleccion.CLICK_UNICO) {
							T[] itemsSeleccionados = getItemsSeleccionados();
							T oldSeleccionado = itemsSeleccionados.length > 0 ? itemsSeleccionados[0] : null;
							removeAllSeleccionSinEventos();
							if (oldSeleccionado != null) {
								onItemDeseleccionado(oldSeleccionado);
							}
							addSeleccion(item);
						} else {
							addSeleccion(item);
						}
					}
					return true;
				}
			}
			return false;
		}

		@Override
		protected boolean onLongClickElemento(AbstractViewHolder<T> viewHolder) {
			T item = viewHolder.getItem();
			if (tipoSeleccion != ETipoSeleccion.NONE) {
				if (!haySeleccionados()) {
					addSeleccion(item);
				} else {
					if (isSeleccionado(item)) {
						removeSeleccion(item);
					} else {
						if (tipoSeleccion == ETipoSeleccion.CLICK_UNICO) {
							removeAllSeleccionSinEventos();
							addSeleccion(item);
						} else {
							addSeleccion(item);
						}
					}
				}
				return true;
			}
			return false;
		}

		@Override
		protected void setItem(T item) {
			T[] itemsSeleccionados = getItemsSeleccionados();
			T old = null;
			for (int i = 0; i < itemsSeleccionados.length && old == null; i++) {
				if (item.getId() == itemsSeleccionados[i].getId()) {
					old = itemsSeleccionados[i];
				}
			}
			if (old != null) {
				removeSeleccion(old);
				addSeleccion(item);
			}
		}

		@Override
		protected void deleteItem(T item) {
			if (isSeleccionado(item)) {
				removeSeleccion(item);
			}
		}

		@Override
		protected void onBindViewHolder(@NonNull AbstractViewHolder<T> holder, int position) {
			T item = listaAdapter.getItem(position);
			boolean seleccionado = isSeleccionado(item);
			holder.setSeleccion(seleccionado);
		}
	}
}

File: listas/src/main/java/com/victorlh/android/framework/listas/adapter/seleccionable/ListaSeleccionableAdapter.kt
package com.victorlh.android.framework.listas.adapter.seleccionable

import com.victorlh.android.framework.listas.adapter.AbstractLista
import com.victorlh.android.framework.listas.adapter.AbstractViewHolder
import com.victorlh.android.framework.listas.adapter.ListaObservable
import com.victorlh.android.framework.listas.adapter.seleccionable.ListaSeleccionable.*
import java.util.*

open class ListaSeleccionableAdapter<T>(override val tipoSeleccion: ETipoSeleccion, val listaAdapter: AbstractLista<T>) : ListaSeleccionable<T> {

    private val observable = ListaSeleccionableObservable()
    private val seleccion = ArrayList<T>()
    private val onStartSeleccionListener: MutableList<OnStartSeleccionListener> = ArrayList()
    private val onStopSeleccionListener: MutableList<OnStopSeleccionListener> = ArrayList()
    private val onSeleccionChangeListeners: MutableList<OnSeleccionChangeListener> = ArrayList()
    private val onItemSeleccionadoListeners: MutableList<OnItemSeleccionadoListener<T>> = ArrayList()
    private val onItemDeseleccionadoListeners: MutableList<OnItemDeseleccionadoListener<T>> = ArrayList()

    init {
        listaAdapter.registrarObservable(observable)
    }

    override fun setSeleccion(items: Array<T>) {
        setSeleccion(listOf(*items))
    }

    override fun setSeleccion(items: List<T>) {
        removeAllSeleccion()
        for (item in items) {
            val id = listaAdapter.getItemId(item)
            val elementoLista = listaAdapter.getItemById(id)
            elementoLista?.let { addSeleccion(it) }
        }
    }

    override fun addSeleccion(item: T) {
        val haySeleccionados = haySeleccionados()
        seleccion.add(item)
        val viewHolder = listaAdapter.getViewHolder(item)
        viewHolder?.seleccionado = true
        if (!haySeleccionados) {
            startSeleccion()
        }
        onItemSeleccionado(item)
        onSeleccionChange()
    }

    override fun removeSeleccion(item: T) {
        seleccion.remove(item)
        val viewHolder = listaAdapter.getViewHolder(item)
        viewHolder?.seleccionado = false
        onItemDeseleccionado(item)
        onSeleccionChange()
        if (!haySeleccionados()) {
            stopSeleccion()
        }
    }

    override fun removeAllSeleccion() {
        seleccion.clear()
        val lista: List<T> = listaAdapter.getLista()
        for (item in lista) {
            val viewHolder = listaAdapter.getViewHolder(item)
            viewHolder?.seleccionado = false
            onItemDeseleccionado(item)
        }
        onSeleccionChange()
        stopSeleccion()
    }

    private fun removeAllSeleccionSinEventos() {
        seleccion.clear()
        val lista: List<T> = listaAdapter.getLista()
        for (item in lista) {
            val viewHolder = listaAdapter.getViewHolder(item)
            viewHolder?.seleccionado = false
        }
    }

    override fun haySeleccionados(): Boolean = seleccion.isNotEmpty()
    override fun isSeleccionado(item: T): Boolean = seleccion.contains(item)

    @Suppress("UNCHECKED_CAST")
    override val itemsSeleccionados: Array<T>
        get() = seleccion.toArray(java.lang.reflect.Array.newInstance(listaAdapter.clazz, 0) as Array<T>)

    override fun addOnStartSeleccionListener(onStartSeleccionListener: OnStartSeleccionListener) {
        this.onStartSeleccionListener.add(onStartSeleccionListener)
    }

    override fun addOnStopSeleccionListener(onStopSeleccionListener: OnStopSeleccionListener) {
        this.onStopSeleccionListener.add(onStopSeleccionListener)
    }

    override fun addOnSeleccionChangeListener(onSeleccionChangeListener: OnSeleccionChangeListener) {
        onSeleccionChangeListeners.add(onSeleccionChangeListener)
    }

    override fun addOnItemDeseleccionadoListener(onItemDeseleccionadoListener: OnItemDeseleccionadoListener<T>) {
        onItemDeseleccionadoListeners.add(onItemDeseleccionadoListener)
    }

    override fun addOnItemSeleccionadoListener(onItemSeleccionadoListener: OnItemSeleccionadoListener<T>) {
        onItemSeleccionadoListeners.add(onItemSeleccionadoListener)
    }

    private fun startSeleccion() {
        for (i in onStartSeleccionListener.indices) {
            val listener = onStartSeleccionListener[i]
            listener.onStartSeleccion()
        }
    }

    private fun stopSeleccion() {
        for (i in onStopSeleccionListener.indices) {
            val listener = onStopSeleccionListener[i]
            listener.onStopSeleccion()
        }
    }

    private fun onSeleccionChange() {
        val size = seleccion.size
        for (i in onSeleccionChangeListeners.indices) {
            val listener = onSeleccionChangeListeners[i]
            listener.onSeleccionChangeListener(size)
        }
    }

    private fun onItemSeleccionado(item: T) {
        for (onItemSeleccionadoListener in onItemSeleccionadoListeners) {
            onItemSeleccionadoListener.onItemSeleccionado(item)
        }
    }

    private fun onItemDeseleccionado(item: T) {
        for (onItemDeseleccionadoListener in onItemDeseleccionadoListeners) {
            onItemDeseleccionadoListener.onItemDeseleccionado(item)
        }
    }

    private inner class ListaSeleccionableObservable : ListaObservable<T>() {
        override fun recargarLista() {
            removeAllSeleccion()
        }

        override fun onClickElemento(viewHolder: AbstractViewHolder<T>): Boolean {
            val item = viewHolder.item
            if (item != null) {
                val haySeleccionados = haySeleccionados()
                if (tipoSeleccion !== ETipoSeleccion.NONE) {
                    if (!haySeleccionados) {
                        if (tipoSeleccion !== ETipoSeleccion.LONGCLICK) {
                            addSeleccion(item)
                            return true
                        }
                    } else {
                        if (isSeleccionado(item)) {
                            removeSeleccion(item)
                        } else {
                            if (tipoSeleccion === ETipoSeleccion.CLICK_UNICO) {
                                val itemsSeleccionados = itemsSeleccionados
                                val oldSeleccionado = if (itemsSeleccionados.isNotEmpty()) itemsSeleccionados[0] else null
                                removeAllSeleccionSinEventos()
                                oldSeleccionado?.let { onItemDeseleccionado(it) }
                                addSeleccion(item)
                            } else {
                                addSeleccion(item)
                            }
                        }
                        return true
                    }
                }
            }
            return false
        }

        override fun onLongClickElemento(viewHolder: AbstractViewHolder<T>): Boolean {
            val item = viewHolder.item
            if (item != null) {
                if (tipoSeleccion !== ETipoSeleccion.NONE) {
                    if (!haySeleccionados()) {
                        addSeleccion(item)
                    } else {
                        if (isSeleccionado(item)) {
                            removeSeleccion(item)
                        } else {
                            if (tipoSeleccion === ETipoSeleccion.CLICK_UNICO) {
                                removeAllSeleccionSinEventos()
                                addSeleccion(item)
                            } else {
                                addSeleccion(item)
                            }
                        }
                    }
                    return true
                }
            }
            return false
        }

        override fun setItem(item: T) {
            val itemsSeleccionados = itemsSeleccionados
            var old: T? = null
            var i = 0
            val itemId = listaAdapter.getItemId(item)
            while (i < itemsSeleccionados.size && old == null) {
                if (itemId == listaAdapter.getItemId(itemsSeleccionados[i])) {
                    old = itemsSeleccionados[i]
                }
                i++
            }
            if (old != null) {
                removeSeleccion(old)
                addSeleccion(item)
            }
        }

        override fun deleteItem(item: T) {
            if (isSeleccionado(item)) {
                removeSeleccion(item)
            }
        }

        override fun onBindViewHolder(holder: AbstractViewHolder<T>, position: Int) {
            val item = listaAdapter.getItem(position)
            val seleccionado = isSeleccionado(item)
            holder.seleccionado = seleccionado
        }
    }

}